internal/provider/resource_tfe_workspace_execution_mode.go

```diff
@@ -0,0 +1,179 @@
+
+	if executionMode != "agent" && agentPoolID != "" {
+		return fmt.Errorf(`execution_mode must be set to "agent" to assign agent_pool_id`)
+	}
+
+	workspaceID := d.Get("workspace_id").(string)
+
+	// Create a new options struct to attach the agent pool to workspace
+	options := tfe.WorkspaceUpdateOptions{
+		AgentPoolID:   tfe.String(agentPoolID),
+		ExecutionMode: tfe.String(executionMode),
+	}
+
+	log.Printf("[DEBUG] Create attachment on workspace with agent pool ID: %s", agentPoolID)
+	workspace, err := config.Client.Workspaces.UpdateByID(ctx, workspaceID, options)
+	if err != nil {
+		return fmt.Errorf("error attaching agent pool ID %s to workspace ID on CREATE %s: %w", agentPoolID, workspaceID, err)
+	}
+
+	d.SetId(workspace.ID)
+	// d.Set() will update the resource state file for execution_mode and agent pool ID
+	d.Set("execution_mode", workspace.ExecutionMode)
+	d.Set("agent_pool_id", workspace.AgentPool.ID)
+
+	return resourceTFEWorkspaceExecutionModeRead(d, meta)
+}
+
+func resourceTFEWorkspaceExecutionModeRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(ConfiguredClient)
+
+	log.Printf("[DEBUG] Read configuration: %s", d.Id())
+	workspace, err := config.Client.Workspaces.ReadByID(ctx, d.Id())
+	if err != nil {
+		if errors.Is(err, tfe.ErrResourceNotFound) {
+			log.Printf("[DEBUG] Workspace %s no longer exists", d.Id())
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("error reading configuration of workspace %s: %w", d.Id(), err)
+	}
+
+	d.Set("workspace_id", workspace.ID)
+
+	var poolID string
+	if workspace.AgentPool != nil {
+		poolID = workspace.AgentPool.ID
+	}
+	d.Set("agent_pool_id", poolID)
+
+	return nil
+}
+
+func resourceTFEWorkspaceExecutionModeUpdate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(ConfiguredClient)
+
+	executionMode := d.Get("execution_mode").(string)
+	agentPoolID := d.Get("agent_pool_id").(string)
+
+	if executionMode == "agent" && agentPoolID == "" {
+		return fmt.Errorf(`agent_pool_id must be provided when execution_mode is "agent"`)
+	}
+
+	if executionMode != "agent" && agentPoolID != "" {
+		return fmt.Errorf(`execution_mode must be set to "agent" to assign agent_pool_id`)
+	}
+
+	workspaceID := d.Get("workspace_id").(string)
+
+	// Create a new options struct to attach the agent pool to workspace
+	options := tfe.WorkspaceUpdateOptions{
+		AgentPoolID:   tfe.String(agentPoolID),
+		ExecutionMode: tfe.String(executionMode),
+	}
+
+	log.Printf("[DEBUG] Create attachment on workspace with agent pool ID: %s", agentPoolID)
+	workspace, err := config.Client.Workspaces.UpdateByID(ctx, workspaceID, options)
+	if err != nil {
+		return fmt.Errorf("error attaching agent pool ID %s to workspace ID %s: %w", agentPoolID, workspaceID, err)
+	}
+
+	d.SetId(workspace.ID)
+	// d.Set() will update state file on tfe_workspace for execution_mode and agent pool ID
+	d.Set("execution_mode", workspace.ExecutionMode)
+	d.Set("agent_pool_id", workspace.AgentPool.ID)
```

Comment:
It's possible for AgentPool to be null, causing a panic if the execution mode is not "agent"

Reply:
I added validation to the customizediff, is that the best place for this check?

Reply:
No, I would just set it to nil if the workspace.AgentPool is nil. There won't always be an AgentPool (like if the execution mode is "remote")